ambry-store/src/main/java/com/github/ambry/store/LogSegment.java

```diff
@@ -572,4 +593,100 @@ public String toString() {
   public StoreMetrics getMetrics() {
     return metrics;
   }
+
+  /**
+   * Interface to run some direct io operations.
+   * @param <T> The return type of the operation.
+   */
+  private interface DirectIOOp<T> {
+    /**
+     * Run the direct io operations, like read, write etc with the provided {@link DirectIoByteChannelAligner}.
+     * @param directIo The direct io file to run the operations.
+     * @return
+     * @throws IOException
+     */
+    T run(DirectIoByteChannelAligner directIo) throws IOException;
+  }
+
+  /**
+   * Direct IO executor interface.
+   */
+  private interface DirectIOExecutor extends Closeable {
+    /**
+     * Execute the {@link DirectIOOp}.
+     * @param op The direct io op to execute.
+     * @param <T> The return type of this operation.
+     * @return
+     * @throws IOException
+     */
+    <T> T execute(DirectIOOp<T> op) throws IOException;
+
+    /**
+     * Flush if there is any buffer.
+     * @throws IOException
+     */
+    void flush() throws IOException;
+
+    @Override
+    void close() throws IOException;
+  }
+
+  /**
+   * A buffered implementation of {@link DirectIOExecutor}.
+   */
+  private class DirectIOBufferedExecutor implements DirectIOExecutor {
+    private DirectIoByteChannelAligner directIo = null;
+    private final int bufferSize;
+
+    public DirectIOBufferedExecutor(int bufferSize) {
+      this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public <T> T execute(DirectIOOp<T> op) throws IOException {
+      // the direct io executes all the operation in the same thread so we don't need
+      // to care able the thread safety.
+      if (directIo == null) {
+        directIo = DirectIoByteChannelAligner.open(file, bufferSize, false);
+      }
+      return op.run(directIo);
```

Comment:
Not necessary in this PR, if we have some metrics to show the how much times execute is called and how many time flush is called, we may see if it helps to accumulate the IO.

Reply:
It's probably not going to be helpful, since when the buffer is full, data will be flushed. But the flush operation is done in the directio library and we don't really know when that happens.